# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [3]:
def addTwo(x):
    return x + 2
print(addTwo(4))

# The lambda version...
# Without using a variable to pass in data
print((lambda x: x + 2)(4))

# Using a variable
a_num = 4
print((lambda x: x + 2)(a_num))

6
6
6


#### Saving to a Variable

In [4]:
f_test = lambda x: x + 2

f_test(4)

6

#### Multiple Inputs

In [7]:
# Multiple inputs with no variable
print((lambda x, y, z: x * y * z)(3, 5, 8))

# Multiple inputs with a variable attached
x_test = lambda x, y, z: x * y * z
print(x_test(3,5,8))

120
120


#### Passing a Lambda into a Function

In [5]:
def multiply(f, num):
    """
        f expects a lambda function
        num expects a number
    """
    return f(num)
multiply(lambda x: x * x, 4) # essentially f = lambda x: x * x and num = 4

16

#### Returning a Lambda from a Function

In [10]:
# Regular defined function
def multiply_tst(num):
    return num * 4

# Function within a function
def returnFunc():
    test = 4
    def multiply(num):
        return num * 2
    return multiply

f_return = returnFunc()
print(returnFunc())
print(f_return(4))

# Lambda function returned from a regular function
def returnLam(b,c):
    return lambda x, a: x + a + b + c
r_lamb = returnLam(4,5)
print(r_lamb(5,5))

<function returnFunc.<locals>.multiply at 0x7fa1a2f8f310>
8
19


#### If Statements within Lambdas

In [11]:
# Lambda x: True if (condition) else False
f_condition = lambda num: num * 2 if num > 10 else num + 2

print(f_condition(8))
print(f_condition(12))
print(f_condition(10))

10
24
12


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [12]:
f = lambda x: x ** 3

print(f(4))

64


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [2]:
# map(func, iterable)
# As iterable had to be something like a list, dict, tuple, set
# normally use map with a pre-defined function but we can use lambdas as well

def squared(num,num2):
    if num < 10 and num2 < 10:
        return num ** 2, num2 ** 2
    else:
        return num, num2
    
numbers = [4, 11, 20, 3, 15, 20]
more_nums = [4, 10, 3, 2, 6]

squared_map_nums = list(map(squared, numbers, more_nums)) # MUST HAVE list( or it will just return that 
print(squared_map_nums)                                   # squared_map_nums is a map object

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### Using Lambda's with Map

In [4]:
# map(lambda x: do something, iterable)
# using lambda and map happens inline (on one line)
print(list(map(lambda x, y: (x ** 2, y ** 2) if x < 10 and y < 10 else (x,y), numbers, more_nums)))


[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [5]:
print(list(map(lambda x, y: (x * 2 - 1, y * 2 - 1), numbers, more_nums)))

[(7, 7), (21, 19), (39, 5), (5, 3), (29, 11)]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [6]:
names = ['Bob', 'Andy', 'Max', 'Evan', 'Angelica']

def a_names(name):
    if name[0].lower() == 'a':
        return True
    else:
        return False

new_names_list = list(filter(a_names, names))
print(new_names_list)

['Andy', 'Angelica']


#### Using Lambda's with Filter()

In [7]:
names_lambda = list(filter(lambda name: True if name[0].lower() == 'a' else False, names))
print(names_lambda)

['Andy', 'Angelica']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [8]:
from statistics import mean
nums = [1, 3, 5, 7, 9, 11, 13, 15]
print(list(filter(lambda num: True if num < mean(nums) else False, nums)))

[1, 3, 5, 7]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [10]:
from functools import reduce
# reduce(function, iterable)
list1 = [2, 4, 6, 8, 10]

def addNums(num1, num2):
    return num1 + num2

result_add = reduce(addNums, list1)
print(result_add)

def subtractNums(num1, num2):
    return num1 - num2

result_sub = reduce(subtractNums, list1)
print(result_sub)

30
-26


#### Using Lambda's with Reduce()

In [12]:
test = reduce(lambda num1, num2: num1 + num2, list1)
print(test)

30


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [14]:
my_list = [1, 2, 3, 4]
print(reduce(lambda x, y: x * y, my_list))

24


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [58]:
def addNums(num):
    # Here is our base case
    if num <= 1:
        print("addNums(1) = 1")
        return num
    else:
        print(f'addNums({num}) = {num} + addNums({num -1})')
        return num + addNums(num -1)

addNums(5)


addNums(5) = 5 + addNums(4)
addNums(4) = 4 + addNums(3)
addNums(3) = 3 + addNums(2)
addNums(2) = 2 + addNums(1)
addNums(1) = 1


15

#### Writing a Factorial Function

In [59]:
# 5! = 5 * 4 * 3 * 2 * 1
def factorial(num):
    if num <=1:
        return 1
    else:
        return num * factorial(num -1)
    
factorial(5)

120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [60]:
def subNums(num):
    if num <=1:
        return num
    else:
        return num - subNums(num-1)

subNums(10)

5

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [104]:
def my_range(stop, start, step = 2):
    while start < stop:
        yield start # yield indicates generator
        start += step

for i in my_range(20, start = 2): # you will get an error if you don't use a for loop
    my_gen_value = i
    print(my_gen_value)

2
4
6
8
10
12
14
16
18


#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [105]:
def numbers_squared(start,stop = 0, step = 1):
    while start >= stop:
        yield start ** 2
        start -= step

for i in numbers_squared(10):
    print(i)

100
81
64
49
36
25
16
9
4
1
0


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [47]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

filt_places = list(filter(lambda place: True if place.strip() else False, places))
print(filt_places)


['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [99]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]
author2 = []
for name in author:
    name = name.title()
    author2.append(name)

print(sorted(author2, key=lambda name: name.split()[-1]))
print(sorted(author, key=lambda name: name.split()[-1]))
# Is there a way to do it without changing the case of the names? I tried and even when I printed 
# name.split()[-1] it would print the last names but then when I added sorted, it would print in a different 
# order (the second print statement - I can't even figure out how that is "sorted")

['Victor Anisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David Hasselhoff']
['Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'Victor aNisimov', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [75]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]
places_2 = []
temps = []
for place in places:
    for item in place:
        if type(item) == str:
            places_2.append(item)
        else:
            temps.append(item)

print(list(map(lambda x, y: (x, (y* (9/5) + 32)), places_2, temps)))

['Nashua', 'Boston', 'Los Angelos', 'Miami']
[32, 12, 44, 29]
[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [69]:
# The fibonacci sequence is actually 0 = 0, 1 = 1, 2 = 1, 3 = 2, 4 = 3, 5 = 5, 6 = 8
# So I think my function is correct but I honestly couldn't figure out how to get the output to be 8 for 5
# But if you put another number in, you get the correct corresponding fib number.
def fibNums(num):
    if num <= 1 or num == 2:
        return 1
    else:
        return fibNums(num - 1) + fibNums(num - 2)
print(fibNums(5))
print(fibNums(6))

5
8
